### DISCUSSION TITLE
Provide ConfigMapping only during BuildStep?

### DISCUSSION THREAD

**Discussion by HerrDerb** [quarkusio/quarkus#43406] [Q&A] [answered]
I would like to provide a ConfigMapping only if a certain BuildStep is applied.
Having the config map defined in the runtimes like:

```java
@ConfigMapping(prefix = "my.configs")
@ConfigRoot(phase = ConfigPhase.BUILD_AND_RUN_TIME_FIXED)
interface MyConfig{

    String property1();
}
```

Having it defined in the runtimes module, this results in this config mapping always being registered and loaded , demanding for the defined property. This property now is feature dependent. If capability A is available during a BuildStep, I would like to add this ConfigMapping to the index. If capability A is not present I will not. I was looking for a fitting BuildItem to provide a ConfigMapping but couldn't find anything fitting.

Can I add a ConfigMapping manually during a BuildStep?

[a comment by an automated account is not shown]

**Reply by radcortez** [reply, marked as answer]
No, currently, there is no way to pick extension mappings selectively. It is possible with application mappings with `@IfBuildProfile` / `@IfBuildProperty`.

An alternative could be to map the feature dependent configuration as `Optionals`.

**Reply by HerrDerb** [reply]
Thanks for clarifications . Would it be worth to create a feature request for a specific BuildItem or is this not that easy implementable ? A workaround which seems to do the job seems to be like this:

There is one way on how to solve it, but I assume that this defaults to a runtime property:
Define the interface without config annotations
```java
interface MyConfig{

    String property1();
}
```
Have a producer for the config bean

```java
public class MyConfigProvider {

    @Produces
    public MyConfig myConfig() {
        SmallRyeConfig config = ConfigProvider.getConfig().unwrap(SmallRyeConfig.class);
        return config.getConfigMapping(MyConfig.class, "herr.derb.application");
    }
}
```
register in the build step like so:
```java
  @BuildStep
    public void additionalBeans(
            BuildProducer<AdditionalBeanBuildItem> additionalBeans,
            BuildProducer<ConfigMappingBuildItem> configMapping) {
        additionalBeans.produce(AdditionalBeanBuildItem.builder()
                .addBeanClass(MyConfigProvider.class)
                .setDefaultScope(DotNames.APPLICATION_SCOPED)
                .build());
        configMapping.produce(new ConfigMappingBuildItem(MyConfig.class, "herr.derb.application"));
     }
```

**Reply by radcortez** [reply]
That shouldn't work. Runtime Mappings require byte code generation for the implementation, triggered by the annotation since we disable the dynamic classloader for runtime.

I prefer to avoid having a build item for this. The correct way to do it is to use `StaticInitConfigBuilderBuildItem` and / or `RunTimeConfigBuilderBuildItem,` which gives you access to the current config builder and customize the configuration to fit your needs. 

The only issue is that the mappings builder while exposed in the builder:
https://github.com/smallrye/smallrye-config/blob/9cc8776b0c51b69e94bb130f9e24cb0072d92010/implementation/src/main/java/io/smallrye/config/SmallRyeConfigBuilder.java#L655-L657

has protected methods:
https://github.com/smallrye/smallrye-config/blob/9cc8776b0c51b69e94bb130f9e24cb0072d92010/implementation/src/main/java/io/smallrye/config/SmallRyeConfigBuilder.java#L764-L799

Feel free to submit a PR to expose them
